addons/web_editor/static/src/js/wysiwyg/widgets/image_crop_widget.js

```diff
@@ -1,174 +1,237 @@
+            this.sessionAttributes.forEach(attr => {
+                this[attr] = this.media.dataset[attr];
+            });
+            return;
         }
-        if (isDBImage) {
-            defs.push(this._rpc({
-                route: '/web_editor/get_image_info',
-                params: params,
-            }).then(function (res) {
-                _.extend(self.imageData, res);
-            }));
+
+        // Get id, mimetype and originalSrc.
+        const {attachment, original} = await this._rpc({
+            route: '/web_editor/get_image_info',
+            params: {src: this.initialSrc.split(/[?#]/)[0]},
+        });
+        if (!attachment) {
+            // Local image that doesn't have an attachment, don't allow crop?
+            // In practice, this can happen if an image is directly linked with its
+            // static url and there is no corresponding attachment, (eg, logo in mass_mailing)
+            this.isCroppable = false;
+            return;
         }
-        return Promise.all(defs);
+        this.originalId = original.id;
+        this.originalSrc = original.image_src;
+        this.originalName = original.name;
+        this.mimetype = original.mimetype;
+        this.attachmentId = attachment.id;
     },
     /**
      * @override
      */
-    start: function () {
-        this.$cropperImage = this.$('.o_cropper_image');
-        if (this.$cropperImage.length) {
-            var data = this.$media.data();
-            var ratio = 0;
-            for (var i = 0; i < this.aspectRatioList.length; i++) {
-                if (this.aspectRatioList[i][1] === data.aspectRatio) {
-                    ratio = this.aspectRatioList[i][2];
-                    break;
-                }
-            }
-            this.$cropperImage.cropper({
-                viewMode: 2,
-                dragMode: 'move',
-                autoCropArea: 1.0,
-                aspectRatio: ratio,
-                data: _.pick(data, 'x', 'y', 'width', 'height', 'rotate', 'scaleX', 'scaleY')
+    async start() {
+        if (!this.isCroppable) {
+            this.displayNotification({
+              type: 'warning',
+              title: _t("This image is an external image"),
+              message: _t("This type of image is not supported for cropping.<br/>If you want to crop it, please first download it from the original source and upload it in Odoo."),
             });
+            return this.destroy();
         }
-        return this._super.apply(this, arguments);
+        const _super = this._super.bind(this);
+        const topLevel = this.document.getElementById('wrapwrap') || this.document;
+        const backdrop = this.document.createElement('div');
+        this.backdrop = backdrop;
+        backdrop.style = `
+            background-color: #8888;
+            position: absolute;
+            top: 0;
+            bottom: 0;
+            left: 0;
+            right: 0;
+            z-index: 1024;
+        `;
```

Comment:
All of this should be done another way:
- Warning and checking for the croppability (yes :)) of the image should not be the role of the widget. Instead, you can create an external function (or a static function maybe) whose role is to instantiate the widget *after checking the cropability*. In the case it is not croppable, the function can throw an error with the message you chose. Then the function caller can decide whether or not it wants to catch that particular error type and if it wants to display a warning in that case.
- What is the `$el` of this widget ? Is this this backdrop? In any case, I feel there should be a template definiton for this widget and use of xml and classes instead of javascript and inline style.

Reply:
recapping from discord for completeness:

The thing is, checking for croppability requires looking up the image in the database. So should I externalized that as well?

> IMO, the widget should only be responsible for the cropping, it is given a croppable image as input and it outputs a cropped image. Even the saving should probably be externalized.

---

Sorry but I'm not following. Taking a croppable image as input and outputting a cropped image is what the cropper library does. This widget checks for croppability, aligns the cropper with the image, and saves the crop data at the end. Basically you're telling me to move all of what this widget does to the the caller?

Reply:
Ok then the question is more simple: why is it a widget if it is only something which calls a library and makes calls to the server?

Reply:
Because it has UI elements that the library doesn't provide (the backdrop and the buttons). I could split this into a Widget for the UI stuff and a bunch of utility functions for loading/storing crop data and attachment data, I'm just not sure why that's desirable.

Reply:
If you use the $el / template properties then it makes sense to use a widget. What made me doubt is that there was none of that when I reviewed or I missed it. Anyways, it seems to be the case now, so let's keep thing that way. It's still not perfect to have a widget/class which instantiates... then only decide to destroy itself if it should not have been instantiated. But let's no waste time on that then.